### PR TITLE
[ios] Fix bug with the "Continue detecting your current location" alert

### PR DESCRIPTION
**Pull request by kirylkaveryn**
Closes https://github.com/organicmaps/organicmaps/issues/6390

### Problem description:
When the device loses GPS for some reason Core raises `pendingTimeoutCallback` which triggers the Alert.
And when it happens in the background mode (when the app is closed ios stops updating location after some time), the alert will become pending in the UI queue and will be shown _when the user opens the application again_. So it produces annoying bugs.

This PR fixes this issue by skipping the alert presenting if the app state is not active.

Because the bug is not easy to reproduce on a real device, it needs to be tested carefully.

Additional improvements:
In the alert, there are 2 options to choose: `Close` and `Continue`.
In Android `Close` stops detecting location and `Continue` do nothing.
https://github.com/organicmaps/organicmaps/blob/f7b526f9d9adf54b940abbceefabf1073500330d/android/app/src/main/java/app/organicmaps/MwmActivity.java#L1994C1-L2017C17

In iOS (for some reason) alert stops detecting location first, and then `Stop` does nothing, and `Continue` restart detection.
https://github.com/organicmaps/organicmaps/blob/394d7d00147b1a47a53dbbf62fe8b19e8a8a8ed3/iphone/Maps/Classes/MapViewController.mm#L523C1-L535C2

So I decided to sync this logic and make it the same way as in Android.

### CODE REVIEW

File: iphone/Maps/Classes/MapViewController.mm [hunk 1]
```diff
@@ -521,6 +521,8 @@ - (void)processMyPositionStateModeEvent:(MWMMyPositionMode)mode {
 }
 
 - (void)processMyPositionPendingTimeout {
+  if (UIApplication.sharedApplication.applicationState != UIApplicationStateActive)
```
Comment (biodranik) [inline iphone/Maps/Classes/MapViewController.mm:524]: Is ignoring the fired timer from the core the best way to fix the issue? What internal state will be there after the timer has fired?
For example, shouldn't the timer be stopped when the app goes into the background?

**Review by biodranik — COMMENTED** [review]
Thanks!

1. Please use `[ios]` tags for iOS-related commits, it helps with release notes.
2. Try to not split the commit's title into two lines with ..., it looks ugly.
3. Did you test these cases/changes? Do they work? 

Does simulator allow a flight mode and unset location as if it's not detected yet?

Flight mode + device reboot + using a fridge, microwave or some other metal objects like foil may help to test it locally.

**Comment by kirylkaveryn** [comment]
> 3\. Did you test these cases/changes? Do they work?
I cannot reproduce the situation when the device loses location updating in BG mode. Because.. it's just don't happens on my device... Even underground or in fridge))
So this is why I need help with testing.

> Does simulator allow a flight mode and unset location as if it's not detected yet?
Nope. The simulator doesn't work in background mode as the device does (push notifications and alerts will not be handled properly in BG mode on simulator).


> Flight mode + device reboot + using a fridge, microwave or some other metal objects like foil may help to test it locally.
It is hard to find stable situation when location fails for me...

File: iphone/Maps/Classes/MapViewController.mm [hunk 2]
```diff
@@ -526,7 +526,7 @@ - (void)processMyPositionPendingTimeout {
     [object processMyPositionPendingTimeout];
```
Comment (biodranik) [inline iphone/Maps/Classes/MapViewController.mm:526]: 0. This callback can be called in the background, right?
1. What is done here?
2. The correct implementation should never call `processMyPositionPendingTimeout` when the app is activated, right? Is there a way to implement it properly?

**Review by biodranik — COMMENTED** [review]
Can you please test it on simulator?
1. Set location to None.
2. Reboot simulator
3. Change `double const kMaxPendingLocationTimeSec = 60.0;` to `5.0` seconds to reproduce issue faster.
4. Start OM and hide it. Wait a bit. Open OM. The issue should reproduce on the master.

FYI: On the master I found that pressing neither Stop nor Continue buttons work in that scenario.

**Comment by kirylkaveryn** [comment]
@biodranik @vng can you please check the new implementatation of the bugfix?
If it will be OK @Jean-BaptisteC - a small fix in android will be needed

Changes:
1. [ios] subscription on `timeoutPending` from framework was removed 
2. [ios] alert `location not found` was removed
3. [drape] When location searching is in 'PendingPosition' (locator is active) and the user taps on the locator, the location stops searching and goes to the `NotFollowNoPosition` state but only when `IsWaitingForLocation`. And location searching will stop.

(on this gif: location not found (airplane mode or etc) -> user disable location -> device found location -> user enable location -> done. (small glitch in gif because I enabled location in the simulator))
Simulator Screen Recording - iPhone 11 Pro - 2024-01-23 at 17 18 48

```swift
void MyPositionController::NextMode(ScreenBase const & screen)
{
  // Skip switching to next mode while we are waiting for position.
  if (IsWaitingForLocation())
  {
    // Stop looking for location.
    m_desiredInitMode = location::Follow;
    ChangeMode(location::NotFollowNoPosition); // < - now mode is changing here
    return;
  }
```
4. [drape] When the app was in Background and entering the Foreground, after 30h and the location state was NotFollowNoPosition (set by user) then the location will continue updating.

```swift
void MyPositionController::OnEnterForeground(double backgroundTime)
{
  if (backgroundTime >= kMaxTimeInBackgroundSec)
  {
      if (m_mode == location::NotFollow)
      {
          ChangeMode(m_isInRouting ? location::FollowAndRotate : location::Follow);
          UpdateViewport(kDoNotChangeZoom);
      }
      else if (m_mode == location::NotFollowNoPosition) // <- New check
      {
          ChangeMode(location::PendingPosition);
          UpdateViewport(kDoNotChangeZoom);
      }
  }
}

```

Location continue detecting and user goes to bg (timeout set from 30h to 8 sec)
Simulator Screen Recording - iPhone 11 Pro - 2024-01-23 at 12 41 10

Location continue detecting and user stop the detecting and goes to bg (timeout set from 30h to 8 sec)
https://github.com/organicmaps/organicmaps/assets/79797627/87bdafad-10ff-4fc5-8e6e-ae12adc0757c

File: drape_frontend/my_position_controller.cpp [hunk 3]
```diff
@@ -342,7 +342,9 @@ void MyPositionController::NextMode(ScreenBase const & screen)
   // Skip switching to next mode while we are waiting for position.
```
Comment (biodranik) [inline drape_frontend/my_position_controller.cpp:342]: This comment should be removed/updated.

File: drape_frontend/my_position_controller.cpp [hunk 4]
```diff
@@ -670,10 +672,18 @@ void MyPositionController::StopLocationFollow()
 
 void MyPositionController::OnEnterForeground(double backgroundTime)
 {
-  if (backgroundTime >= kMaxTimeInBackgroundSec && m_mode == location::NotFollow)
+  if (backgroundTime >= kMaxTimeInBackgroundSec)
   {
-    ChangeMode(m_isInRouting ? location::FollowAndRotate : location::Follow);
-    UpdateViewport(kDoNotChangeZoom);
+      if (m_mode == location::NotFollow)
```
Comment (biodranik) [inline drape_frontend/my_position_controller.cpp:677]: We're using 2 spaces as indents.

**Review by biodranik — COMMENTED** [review]
@vng can you please take a look?

**Comment by biodranik** [comment]
1. Apply this PR and yours #7193
2. Set flight mode on the device
3. Reboot
4. Launch debug OM version.
5. Observe that the location is NOT detected at all, despite waiting, pressing the radar button several times, etc.
6. Observe that the screen turns off when the location is searched for (that stops location services, right? IMO it would be better to keep the screen on while location is searched for)
7. Observe that other apps detect location in several seconds properly.
8. After location was detected in other apps, switch to OM and observe that location is NOT detected, despite pressing location button.
9. Restart OM and observe that location is detected correctly.

I assume that location manager isn't started properly, or both pull requests may interfere with each other.

File: drape_frontend/my_position_controller.cpp (hunk 4 above)
Comment (kirylkaveryn) [reply drape_frontend/my_position_controller.cpp:677]: Fixed!

File: drape_frontend/my_position_controller.cpp (hunk 3 above)
Comment (kirylkaveryn) [reply drape_frontend/my_position_controller.cpp]: Fixed!

**Review by vng — APPROVED** [review]
Should we make any changes in Android or can merge it now?

**Comment by vng** [comment]
Merge https://github.com/organicmaps/organicmaps/pull/7193 first!

File: drape_frontend/my_position_controller.cpp [hunk 5]
```diff
@@ -339,10 +339,11 @@ void MyPositionController::ResetRenderShape()
 
 void MyPositionController::NextMode(ScreenBase const & screen)
 {
-  // Skip switching to next mode while we are waiting for position.
   if (IsWaitingForLocation())
   {
+    // Stop looking for location.
     m_desiredInitMode = location::Follow;
+    ChangeMode(location::NotFollowNoPosition);
```
[the remark this answers is not in the document]
Comment (vng) [reply drape_frontend/my_position_controller.cpp:346]: Manualy Turn Off the location when pressing the location button in the PendingPosition state.

File: drape_frontend/my_position_controller.cpp [hunk 6]
```diff
@@ -670,10 +671,17 @@ void MyPositionController::StopLocationFollow()
...
 void MyPositionController::OnEnterForeground(double backgroundTime)
 {
-  if (backgroundTime >= kMaxTimeInBackgroundSec && m_mode == location::NotFollow)
+  if (backgroundTime >= kMaxTimeInBackgroundSec)
   {
-    ChangeMode(m_isInRouting ? location::FollowAndRotate : location::Follow);
-    UpdateViewport(kDoNotChangeZoom);
+    if (m_mode == location::NotFollow)
+    {
+      ChangeMode(m_isInRouting ? location::FollowAndRotate : location::Follow);
```
[the remark this answers is not in the document]
Comment (vng) [reply drape_frontend/my_position_controller.cpp:678]: It was before. Automatically start looking for the location after some timeout (30 hours)

**Comment by kirylkaveryn** [comment]
@rtsisyk I added some comments to better understand what happens.

Main points:
1.   When the app is awaiting location (indicator is active) and the user presses on the indicator, location updates will be stopped and go into the NotFollowNoPosition state. The next press on the indicator will start location updates again.
2. If the app goes from background into foreground state, 30h was left and the user stopped location manually during the last session, then the app will try to find the location without following (because the user disabled location updates).

**Pull request merged**